### DISCUSSION TITLE
Consistent Issue getting HTTP Header using `preRenderView`

### DISCUSSION THREAD

**Discussion by NCIDevRonR** [primefaces/community#2888] [PrimeFaces] [answered]
In our application, we go through a pre-render step, get a subjectDN value from the header when our user opens the app, and we authenticate the user from there, then navigate to the first application page.  From the day I first inherited our codebase in PrimeFaces version 6.2 through our current updates in PF version 14, we've had this issue where if the user navigates super-fast to our page, they see our error screen.  If they reload the page right after, they see our main screen.  [I've reproduced our issue here](https://github.com/NCIDevRonR/primefaces-test).  What can I do to resolve this?

**Comment by melloware** [comment]
1. PreRenderView was the old way you might want to try ViewAction instead see BalusC's explanation: https://stackoverflow.com/a/9844616/502366    or a better one: https://stackoverflow.com/a/6377957/502366

2. What do you mean by navigates super fast?

**Reply by NCIDevRonR** [reply]
I mean if I set up the same header in the browser that the filter expects, I can get different results depending on how soon after I open my browser I attempt to navigate to the test app.  If I wait a few seconds, I get the welcome screen.  If I navigate as soon as the browser comes up, I get the error screen.

**Reply by melloware** [reply]
have you debugged the Filter with log lines to see whether the header is set or not?  The Filter in theory should RUN before the JSF code in your `preRenderView` (in theory).

**Reply by NCIDevRonR** [reply]
I've just added a System.out to check subjectDN for null just before the if, and I see what I expected to see.  When I navigate quickly to the page, there's a null in subjectDN and I see the error page, and when I wait a few seconds before navigating onto the page, subjectDN is populated, and I see the welcome page.

**Reply by NCIDevRonR** [reply]
I have updated my test app with a ViewAction, and I've set it up to retry 15 times if the subjectDN handshake fails.  The original issue still exits. If the user navigates immediately upon opening the browser, the back-end tries 15 times unsuccessfully to handshake with subjectDN. If the user waits five seconds after opening the browser and then navigates to the test page, the back-end succeeds in handshaking with subjectDN the first time.  Is there a way to rewrite the retry loop, to maximize the likelihood of a successful retry?

**Reply by melloware** [reply]
not sure i still can't explain why that is even happening.

**Reply by melloware** [reply]
@NCIDevRonR did you ever figure this out?

**Reply by NCIDevRonR** [reply, marked as answer]
Hey, it's been a WHILE!!  :)  I don't think my team and I ever cracked this one.  I think we just left it as a "known issue".  We're under a directive to migrate to another platform anyway.  If you want to go ahead and close it, that's fine by me.
